actors/evm/src/interpreter/precompiles/bls12_381/g1_msm.rs

```diff
@@ -0,0 +1,277 @@
+use crate::interpreter::{
+    System,
+    precompiles::{
+        PrecompileContext, PrecompileError, PrecompileResult,
+        bls_util::{SCALAR_LENGTH_BITS, is_infinity, p1_scalar_mul, p1_to_affine},
+    },
+};
+use fil_actors_runtime::runtime::Runtime;
+
+use crate::interpreter::precompiles::bls_util::{
+    G1_MSM_INPUT_LENGTH, G1_OUTPUT_LENGTH, PADDED_G1_LENGTH, encode_g1_point, extract_g1_input,
+    read_scalar,
+};
+
+use blst::{MultiPoint, blst_p1_affine, blst_scalar};
+
+/// **BLS12_G1MSM Precompile**
+///
+/// Implements G1 multi-scalar multiplication (MSM) according to [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537#abi-for-g1-multiexponentiation).
+/// The input is expected to be `160 * k` bytes, where each 160-byte segment consists of:
+/// - A 128-byte encoding of a G1 point (padded, in affine form)
+/// - A 32-byte encoding of a scalar value in Big-Endian
+///
+/// The output is an encoding of a single G1 point (128 bytes).
+pub fn bls12_g1msm<RT: Runtime>(
+    _: &mut System<RT>,
+    input: &[u8],
+    _: PrecompileContext,
+) -> PrecompileResult {
+    // Validate that the input length is non-zero and an exact multiple of the per-pair length.
+    let input_len = input.len();
+    if input_len == 0 || input_len % G1_MSM_INPUT_LENGTH != 0 {
+        return Err(PrecompileError::IncorrectInputSize);
+    }
+
+    // Determine the number of (point, scalar) pairs.
+    let k = input_len / G1_MSM_INPUT_LENGTH;
+    let mut g1_points: Vec<blst_p1_affine> = Vec::with_capacity(k);
+    let mut scalars: Vec<blst_scalar> = Vec::with_capacity(k);
+
+    // Process each (point, scalar) pair.
+    for i in 0..k {
+        let offset = i * G1_MSM_INPUT_LENGTH;
+        let encoded_point = &input[offset..offset + PADDED_G1_LENGTH];
+        let encoded_scalar = &input[offset + PADDED_G1_LENGTH..offset + G1_MSM_INPUT_LENGTH];
```

Comment:
Normally I'd recommend using the `ValueReader` but... that's specifically for precompiles that treat the input as having "infinite trailing zeros". And, apparently, these new precompiles don't do that.